File: ruby/3.rb
# time ruby 3.rb
# 6857
# 0.48user 0.02system 0:00.52elapsed 95%CPU (0avgtext+0avgdata 0maxresident)k
# 120inputs+0outputs (1major+1235minor)pagefaults 0swaps

def fermat_factors(n)
  return [2, n/2] if n % 2 == 0
  a = Math.sqrt(n).ceil
  b = a*a-n
  s = Math.sqrt(b)
  while s != s.round
    a += 1
    b = a*a-n
    s = Math.sqrt(b)
  end
  [a-s, a+s]
end

def prime_factors(n)
  factors = fermat_factors(n)
  if factors.include? 1
    factors.delete_if { |f| f == 1 }
  else
    factors.map { |f| prime_factors(f) }.inject(&:concat)
  end
end

puts prime_factors(600851475143).max.round

File: scala/3.scala
// JAVA_OPTS="-Xmx1024M" time scala -deprecation 3.scala
// 6857
// 1.48user 0.17system 0:04.06elapsed 40%CPU (0avgtext+0avgdata 0maxresident)k
// 0inputs+0outputs (0major+11382minor)pagefaults 0swaps

import scala.math._

def fermatFactors(n: Long): Set[Long] = {
  if (n % 2 == 0) return Set(2, n / 2)
  var a = ceil(sqrt(n))
  var b = a * a - n
  var s = sqrt(b)
  while (s != round(s)) {
    a += 1
    b = a * a -n
    s = sqrt(b)
  }
  Set(a - s, a + s).map(_.toLong)
}

def primeFactors(n: Long): Set[Long] = {
  val factors = fermatFactors(n)
  if (factors contains 1)
    factors - 1
  else
    factors.map(f => primeFactors(f)).reduceLeft(_ ++ _)
}

println(primeFactors(600851475143L).reduceLeft(_ max _))
